Add Cargo team charter. (#12010)
rust-lang/cargo: src/doc/contrib/src/team.md

# Cargo Team

## Mission

The Cargo Team is a group of volunteers that support the Rust community in developing and maintaining Cargo, the Rust package manager and build tool.
The team is responsible for deciding how Cargo and its related libraries operate and evolve.
The team has a shared responsibility with the [crates.io team] for the design and usage of Cargo's index format and its registry API as it relates to the [crates.io] service.

The team is expected to keep Cargo in an operational state, to support Rust's 6-week release cycle, and to uphold the [Design Principles] of Cargo.

[crates.io team]: https://www.rust-lang.org/governance/teams/crates-io
[crates.io]: https://crates.io/
[Design Principles]: design.md

## Team membership

The Cargo Team consists of team members with one serving as a team leader.
The team leader is responsible for coordinating the team and providing a contact point with other teams.

> The leader is selected by a vote of the existing members.

Reviewer: ```suggestion
The leader is selected by consensus of the existing members with no objections.
```

Team members are expected to regularly participate in at least some of the following membership-related activities.
Members are not expected to participate in all of these activities, but exhibit some interest and involvement in the project that covers some of these activities.

- Attending meetings
- Reviewing contributions (auto-assignment is managed in [triagebot.toml])
- Triaging and responding to issues
- Mentoring new contributors
- Shepherding major changes and RFCs
- Coordinating interaction with other Rust groups and outside interests
- Managing and updating the policies of the Cargo Team itself
- Keeping up with maintenance of the Cargo codebase, ensuring it stays functional and that infrastructure and team processes continue to run smoothly

Breaks and vacations are welcome and encouraged.
If a member is no longer participating after a few months, they may be asked to step down.

Members are required to always:

- Represent the Rust project in a way that upholds the [Rust code of conduct][coc] to a high standard.
- Represent the Cargo Team in a way that upholds the expectations of this charter, and be friendly, welcoming, and constructive with contributors and users.

> Members are given privileges, such as:

Reviewer 2: One thing missing here is the permission of crates.io for crates maintained by the Cargo team. Who has the right to publish manually? Or should we make all things in automation even for a point release of cargo-the-library?

Reviewer 3: If we adopt cargo-release, it would semi-automate the process. One of the more recent features is you can set a workspace-wide cargo-owner and it will be set when publishing new crates. I added this when I I realized there were newer clap crates without the team added to them.

Author: At this time I don't believe the infra team is willing to support tokens for auto-publishing from CI.  They are at least waiting on scoped tokens. 

I'm not sure how best to manage crates.io ownership. I think for now people can be added on an as-needed basis. I think we should preferably keep the ownership list small, perhaps eventually with no individual owners.  I'm not sure if we can fully automate it once scoped tokens are available (such as figuring out how to trigger releases, etc.).

Reviewer: > At this time I don't believe the infra team is willing to support tokens for auto-publishing from CI. They are at least waiting on scoped tokens.

Experimental support for scoped tokens has shipped, so it's possible to do this now.

- Merge permissions (bors rights)
- Issue and project management (GitHub permissions)
- Voting and decision making (RFCs, major changes)
- Access to private communications related to team management and security discussions

[coc]: https://www.rust-lang.org/policies/code-of-conduct
[triagebot.toml]: https://github.com/rust-lang/cargo/blob/master/triagebot.toml

### Meetings

The team meets on a weekly basis on a video chat.
If you are interested in participating, feel free to contact us on [Zulip].

### Becoming a member

A contributor can become a member of the Cargo Team by requesting a review or being nominated by one of the existing members.

> They can be added by a unanimous vote from the team.

Reviewer: ```suggestion
They can be added by unanimous consent of the team.
The team lead or another member of the team will also confirm with the moderation team that there are no concerns involving the proposed team member.
```

# Cargo Team

## Mission

The Cargo Team is a group of volunteers that support the Rust community in developing and maintaining Cargo, the Rust package manager and build tool.
The team is responsible for deciding how Cargo and its related libraries operate and evolve.
The team has a shared responsibility with the [crates.io team] for the design and usage of Cargo's index format and its registry API as it relates to the [crates.io] service.

The team is expected to keep Cargo in an operational state, to support Rust's 6-week release cycle, and to uphold the [Design Principles] of Cargo.

[crates.io team]: https://www.rust-lang.org/governance/teams/crates-io
[crates.io]: https://crates.io/
[Design Principles]: design.md

## Team membership

> The Cargo Team consists of team members, and a team leader.

Author: Sounds good, I merged the sections.

Membership is maintained in the [Rust team database].

[Rust team database]: https://github.com/rust-lang/team/blob/master/teams/cargo.toml

### Membership expectations

Team members are expected to regularly participate in the following membership-related activities:

- Voting on RFCs and major change proposals
- Attending meetings

> - Reviewing contributions

Author: Add a link.

- Triaging and responding to issues
- Mentoring new contributors
- Shepherding major changes and RFCs
- Coordinating interaction with other Rust groups and outside interests
- Managing and updating the policies of the Cargo Team itself
- Keeping up with maintenance of the Cargo codebase, ensuring it stays functional and that infrastructure and team processes continue to run smoothly

> Members are not expected to participate in all of these activities, but exhibit some interest and involvement in the project that covers some of these activities.

Author: I split the list up and tried to make it clearer up-front that these aren't all requirements.  I wasn't quite sure how to structure it, so let me know if you have any ideas to clean it up.

Voting on RFCs and major change proposals is expected from all members.
Breaks and vacations are welcome and encouraged.
If a member is no longer participating after a few months, they may be asked to step down.

Members are required to always:

- Represent the Rust project in a way that upholds the [Rust code of conduct][coc] to a high standard.

> - Represent the Cargo Team in a way that upholds the expectations of this charter, and be friendly, welcoming, and constructive with contributors and users.

Author: Can you say more about what your thoughts are here? I think I have a vague idea of what you're getting at here, but I'm not sure how to put it into words. I assume this is trying to make the distinction of making some public statement, and wanting to distinguish between "I'm speaking for the whole team" and "I'm just giving my personal viewpoint"?

> - Represent the Cargo Team in a way that upholds the expectations of this charter, and be friendly, welcoming, and constructive with contributors and users.

Reviewer 3: Thats it. I'm not sure what to even put down for a policy, if anything. Its just something I try to keep in mind

> - Represent the Cargo Team in a way that upholds the expectations of this charter, and be friendly, welcoming, and constructive with contributors and users.

Reviewer 2: We would always be recognized as “Cargo team member” or whatever your strongest public image when interacting with the community. That's a thing one can never shave off even after stepping down. 

For example, when one team member talks about `cargo-eval` pre-RFC, even when they state it's their own personal opinion. They still have a vote to affect the decision. People then tend to see it as a portion of the team's opinion if they find the connection.

That's an inevitable consequence of being a formal member.

> - Represent the Cargo Team in a way that upholds the expectations of this charter, and be friendly, welcoming, and constructive with contributors and users.

Reviewer 2: In terms of making a statement for the whole team, we usually start like “after the team's recent discussion, we….” I think that kind of speaking gives enough information to the audience to decide where it comes from. 

Just be clear. I am fine with not having a policy here for now.

> - Represent the Cargo Team in a way that upholds the expectations of this charter, and be friendly, welcoming, and constructive with contributors and users.

Reviewer: It's absolutely the case that we'll be perceived as Cargo team members, but with some care we can make sure to state when we're speaking for the team, and when we're specifically *not* speaking for the team.

> - Represent the Cargo Team in a way that upholds the expectations of this charter, and be friendly, welcoming, and constructive with contributors and users.

Reviewer 4: Following today's episode of the rust drama... being unclear about whether someone was speaking on behalf of the team or just expressing their own opinion may have been a key contributor to this round of drama.

Members are given privileges, such as:

- Merge permissions (bors rights)
- Issue and project management (GitHub permissions)
- Voting and decision making (RFCs, major changes)
- Access to private communications related to team management and security discussions

[coc]: https://www.rust-lang.org/policies/code-of-conduct

### Meetings

The team meets on a weekly basis on a video chat.
If you are interested in participating, feel free to contact us on [Zulip].

### Becoming a member

A contributor can become a member of the Cargo Team by requesting a review or being nominated by one of the existing members.
They can be added by a unanimous vote from the team.

Contributors who wish to join the team should exhibit an interest in carrying the design principles of Cargo and participate in some of the activities listed above in [Membership Expectations](#membership-expectations).

Members may leave at any time, preferably by letting the team know ahead of time.

### Leadership

> The team leader is responsible for coordinating the team and providing a contact point with other teams.

Author: Are there any specific details that you are thinking of? The only specific thing I can imagine is maybe notifications of security issues, but that's kind of the purview of the security team, and is already somewhat covered by being "a contact point with other teams".  

If there are specific things you'd like to see here, let me know.

> The team leader is responsible for coordinating the team and providing a contact point with other teams.

Reviewer 2: It's more like “if you were about to step down today, what would you want the new team lead to know.” 

I don't have any specific idea on that, just feel like knowing more details can make the team more robust. If there is nothing new that's fine!

> The team leader is responsible for coordinating the team and providing a contact point with other teams.

Author: I can't think of too many specific things.
I assume the next leader would also be added as a member to the devtools team. Right now that doesn't really involve much other than monitoring the devtools zulip channel.

  - Bug fixes that do not introduce backwards-incompatible changes, and adhere to Cargo's expected behavior.
  - Addition of new warnings, or other diagnostic changes.
  - New or updated documentation.
  - Localized refactorings (that is, those that do not have a significant, wide-ranging impact to the usage and maintenance of the codebase).
  - Minor or planned changes to Cargo's console output.
  - Beta backports that clearly address a regression, and are expected to be low-risk.
  - Development of a previously approved unstable feature that matches the expected development of that feature.

- Small features or changes, large refactorings, or major changes to Cargo's codebase or process require an approval by the team via consensus.
  These decisions can be done via the FCP process of [rfcbot], or in an ad-hoc manner such as during a team meeting.
  rfcbot FCP requests do not require waiting for the 10-day feedback window if there is a complete team consensus, as this process is mostly aimed at polling the team rather than publicly soliciting feedback.
  Though, public feedback is welcome at any time.

  Some examples of what this might cover are:

>   - Addition of a new, minor command-line argument, or an addition of an option to an existing one.

Reviewer 2: I wonder which category adding a new dependency falls into. We've discussed this from time to time. Is it worth a mention?

Author: That's a good question. I would kind of like to make that a separate thing, outside of this chapter and PR.  I'm not sure where it would go, but I think it would be great to have written guidelines for dependencies. I'm wondering if there are other guidelines that are worth writing down? Like maybe review guidelines? Test requirements? Guidelines for when to revert?

    These decisions are usually processed via private channels by the entirety of the team.
  - A change that is a "one-way door".
    That is, something that is difficult to reverse without breaking backwards compatibility.

- Larger features should usually go through the [RFC process].
  This usually involves first soliciting feedback from the Cargo team and the rest of the community, often via the [Rust Internals] discussion board, [Cargo's issue tracker], and the [Zulip] channel.
  If there is positive feedback to the idea, the next step is to formally post an RFC on the RFC repo.
  The community and the Cargo team will have an opportunity to provide feedback on the proposal.
  After some period of time, the Cargo team may decide to either accept, postpone, or close a proposal based on the interest in the proposal and the team's availability.

  Some examples of what this might cover are:
  - Major changes or new features or options in `Cargo.toml` or the config files.
  - Changes to the registry index or API.
  - New or changed CLI options that are expected to have a significant impact on how Cargo is used.

>   - New `cargo` commands.

Reviewer 3: We have added some commands with a change proposal, rather than an FCP.  Should we make a distinction on this?

I'm assuming RFC is more needed when the command has non-trivial workflows and hasn't been broadly adopted by the community to vet those workflows. 
- `cargo rm` is more trivial
- `cargo add` is a little less trivial but had broad use for a long time
- `cargo-script` has non-trivial workflows and isn't pervasive.  Even if it was, the workflows are non-trivial enough to warrant an RFC to re-evaluate those design decisions
  - I think `cargo-nextest` is a good example of this also.  It has done a lot to improve workflows but in a way that works around limitations in the ecosystem.  Instead of replacing `cargo test` with it, we'd need to decide what non-workaround solutions to the problems look like

Author: Yea, this is a bit of a gray area.  I added some extra text here, though I'm not sure if it is really complete. Let me know if you have any suggestions to the wording.

  Such experiments are expected to be nightly-only (see the [Unstable] chapter), and involve efforts to shape the final result via exploration, testing, and public involvement.
  Any such features *must* ultimately have an RFC approved before they can be stabilized.

[rfcbot]: https://github.com/rust-lang/rfcbot-rs
[RFC process]: https://github.com/rust-lang/rfcs/
[Rust Internals]: https://internals.rust-lang.org/
[Unstable]: process/unstable.md
[Backwards compatibility]: design.md#backwards-compatibility
[Stable backports]: process/release.md#stable-backports
[SemVer chapter]: https://doc.rust-lang.org/cargo/reference/semver.html

## Contacting the team

The team may be contacted through several channels:

- If you have a **security concern**, please refer to Rust's [security policy] for the correct contact method.
- Issues and feature requests can be submitted to [Cargo's issue tracker].
  Please see the [Issues chapter] for more detail.
- The [`t-cargo` Zulip channel][Zulip] stream is the chat platform the Cargo Team uses to coordinate on.

> - The `cargo@rust-lang.org` email address can be used to contact the team.

Author: I intentionally wasn't wanting to make it a link, just to avoid the temptation to actually click it, but that is a bad motivation.
